src/System.Private.CoreLib/src/System/Threading/LowLevelSpinWaiter.cs

```diff
@@ -20,24 +20,13 @@ internal struct LowLevelSpinWaiter
         private const int SpinCount = 8;
         private const int SpinSleep0Threshold = 4;
 
-        private static int s_processorCount;
-
         private int _spinningThreadCount;
 
-        public void Initialize()
-        {
-            if (s_processorCount == 0)
-            {
-                s_processorCount = Environment.ProcessorCount;
-            }
-        }
-
         public bool SpinWaitForCondition(Func<bool> condition)
         {
             Debug.Assert(condition != null);
-            Debug.Assert(s_processorCount > 0);
 
-            int processorCount = s_processorCount;
+            int processorCount = PlatformHelper.ProcessorCount;
```

Comment:
I'm not sure that this would work because `PlatformHelper` has a static constructor, and this path may be taken during lazy class construction, and may create a circular dependency. Last I remember there wasn't an obvious error and those type of issues were difficult to diagnose. I'm not sure actually if it's even ok to use `Environment.ProcessorCount` as it was doing before in `Initialize`, as `Environment` too has a static constructor. It may be necessary to call into the runtime directly, but I'll have to follow up on that. For now I think this is probably no worse than before.

Reply:
Getting rid of static constructor in `PlatformHelper` should be a matter of changing:
```internal static readonly bool IsSingleProcessor = ProcessorCount == 1;```
to
```internal static bool IsSingleProcessor => ProcessorCount == 1;```
...but that requires some more analysis of the possible consequences.

Reply:
I think it worked before because the `WaitSubsystem` has eager static construction attribute and a static `LowLevelLock` instance, the constructor of which would initialize the rest of the types used on that path including `Environment` eagerly.

We could remove the static constructor from `PlatformHelper` but I'd like to avoid expanding the scope of that issue, as with that changes in the future have to be careful to not introduce a static constructor to PlatformHelper. I would be good to contain that issue in this type.

I think what I suggested before is a decent approach for now - to make the `Initialize` method that was there before static and call it from the constructors of `LowLevelLock` and `LowLevelLifoSemaphore`.

As for `IsSingleProcessor`, that typically doesn't change but for spin-waits ideally in addition to the proc count we would also consider the processor affinity, which can change. Anyway that stuff can be fixed later when necessary.

Reply:
Just wondering, wouldn't it work to stick

```c#
#if CORERT
[EagerStaticClassConstruction]
#endif
```

in front of `PlatformHelper`?

Reply:
It might be neatest to just have `private static int s_processorCount = Environment.ProcessorCount` and put an eager static construction attribute on this type, but again I'm not sure why that wasn't done in the first place, there's probably something else wrong with that

Reply:
Yep that could work. I'm curious what broke though, maybe it has nothing to do with this. If adding that fixes it maybe that can be a solution.

Reply:
I actually expected that the underlying issue could go unnoticed because that's kinda what happened to me before

Reply:
I'll try couple of the solutions...

> I think it worked before because the `WaitSubsystem` has eager static construction attribute and a static `LowLevelLock` instance, the constructor of which would initialize the rest of the types used on that path including `Environment` eagerly.

Makes sense. Thanks for analyzing it.

> We could remove the static constructor from `PlatformHelper` but I'd like to avoid expanding the scope of that issue, as with that changes in the future have to be careful to not introduce a static constructor to PlatformHelper. It would be good to contain that issue in this type.

Yeah, it would be nice to contain it. I think `PlatformHelper` is small enough that it would make sense to try to contain it there (if there's easy way).

> I think what I suggested before is a decent approach for now - to make the `Initialize` method that was there before static and call it from the constructors of `LowLevelLock` and `LowLevelLifoSemaphore`.

It's not actually necessary for `LowLevelLifoSemaphore` since it's used only from thread pool and much later. `PlatformHelper` works just fine there.

Reply:
@jkotas, can you advise on what might be a reasonable solution here?

This type (`LowLevelSpinWaiter`) is used by `LowLevelLock`, in turn used by the wait subsystem, in turn used by lazy class construction. IIRC then paths of `LowLevelSpinWaiter` that may be taken during lazy class construction cannot reference other types with static constructors, is that right? Would it be appropriate to use `[EagerStaticClassConstruction]` on dependency types in those paths that have static constructors to work around an issue there?

Reply:
Previously I believe it was working because during an eager static construction (for `WaitSubsystem`), the statics necessary for this type were initialized

Reply:
The rule for `[EagerStaticClassConstruction]` is that you need to have tight control and understanding over all code that is run from the static constructor of the type, and it is best to not call to other types with static constructors (or that are big and complex like Environment and somebody may add random static constructor to them).

Reply:
> can you advise on what might be a reasonable solution here?

Call `Runtime.RuntimeImports.RhGetProcessCpuCount()` directly from PlatformHelper. Do not go via Environment.

Reply:
Looks like a can of worms I didn't want to open :-/ I'll be more than happy to address it in separate PR, but right now I am only burning CPU time of the CI machines hopelessly trying to fix something that was not a nice code but at least worked somehow.

Reply:
> I'm curious what broke though

The thing that actually broke is that I accidentally removed `spinCount++` line from `LowLevelSemaphore` while doing the code edits. Kinda stupid mistake...

Reply:
Thanks @jkotas, that sounds fine to me @filipnavara

Reply:
> accidentally removed `spinCount++`

I see, makes sense

Reply:
`PlatformHelper` is in the shared partition and that implementation wouldn't work for CoreCLR. It could be unshared, but I think an easy and reasonable enough solution for now may be:

`LowLevelSpinWaiter`:

```c#
private static int s_processorCount;
```

`LowLevelSpinWaiter::SpinWaitForCondition`:

```c#
...
int processorCount = s_processorCount;
if (processorCount == 0)
{
    s_processorCount = processorCount = Runtime.RuntimeImports.RhGetProcessCpuCount();
}
...
```

Reply:
Or even just forget the static variable and just call `Runtime.RuntimeImports.RhGetProcessCpuCount()` every time, this is already the slow path and would be fine perf-wise

Reply:
I would not exactly call it a slow path, but it's probably ok either way.

Reply:
Adding a bit of work to the spin paths of a lock has not mattered in a noticeable way in the past. It would be fine to preinitialize the static variable with the proc count to avoid that extra work. Issues with changing proc counts, which are there in multiple places, can be dealt with separately.